### ISSUE TITLE
is `curl --insecure` still needed ?

### ISSUE THREAD

**Issue by AltGr** [ocaml/opam#2006]
This was added because of compatibility issues on many common installations, in which curl didn't have access to a reasonable list of root certificates (in particular for github). I am wondering if this is still an issue now ?

We still double-check md5s of course.

**Comment by avsm** [comment]
The other point of note is that Cohttp is now in good enough shape to download the vast majority of URLs via the safe TLS stack.  I'm testing this out specifically in the https://github.com/avsm/opam-mirror utility

**Comment by AltGr** [comment]
Not relying on Curl anymore would indeed be quite cool.

**Comment by avsm** [comment]
Frustratingly, Curl now fails on MacOS X on some sites:

```
curl "--write-out" "%{http_code}\\n" "--insecure" "--retry" "3" "--retry-delay" "2" "-OL" "https://ocaml.janestreet.com/ocaml-core/112.17/files/sexplib-112.17.00.tar.gz"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0000
curl: (56) SSLRead() return error -9841
```

This appears to be due to a `curl` configuration option: http://stackoverflow.com/questions/26461966/osx-10-10-curl-post-to-https-url-gives-sslread-error

@yminsky, did the JS site recently disable SSLv3 to trigger this?  I can workaround it by removing the `https` from the distfiles for now, and hope the md5 checksum is strong enough.

**Comment by dsheets** [comment]
`--insecure` already meant that we had no authentication against CA bundle...

**Comment by avsm** [comment]
The error above was due to protocol negotiation and not CA checking, I think

**Comment by AltGr** [comment]
Travis reports such failures to sourceforge, too (on OSX)
see for example 
https://travis-ci.org/ocaml/opam/jobs/51460132

**Comment by camlspotter** [comment]
Is there any workaround for this issue?

**Comment by AltGr** [comment]
the latter ? Hm, if it doesn't find `curl` in your `PATH`, OPAM will use `wget` instead.
If you have a working `curl` command somewhere, set `OPAMCURL` to point at it (there is no similar way to specify a `wget` command at the moment though, curl-specific command-line args will be used after `$OPAMCURL`).

**Comment by avsm** [comment]
The issue is that curl is broken on MacOS x, so we need to force the use of wget. Time for another conf variable to OPAMFETCH or something? Could also add the cohttp-curl binary as an option too

**Comment by AltGr** [comment]
The issue at hand (curl on OSX) is not really the one from the OP, but on that:
I confirm that it'll be fixed by #2040 -- as long that you have wget installed (if you only have curl, OPAM will still try to use it).

It doesn't only choose wget by default on the mac, I made it fully customisable: you can specify the command using OPAMFETCH or the `download-command` in ~/.opam/config ; variable expansion allows `url` and `out` in the command, e.g. in ~/.opam/config:
```
download-command: [ "cohttp-curl-lwt" url "-o" out ]
```
Or, equivalent: `OPAMFETCH="cohttp-curl-lwt %{url}% -o %{out}%`

Of course if you do that and the exe is installed within OPAM, you may get in trouble if it gets removed at some point or you create a new switch :)

While I was at it there is a similar interface for the external solver now.

**Comment by UnixJunkie** [comment]
maybe we should sign all packages with something stronger than MD5, also.

**Comment by AltGr** [comment]
The roadmap for 1.3 has end-to-end signatures and signed consistent repository snapshots :)
We're settling on the design and will open it for review soon.

**Comment by hannesm** [comment]
to be picky "sign [...] stronger than md5": MD5 does not sign anything, it is a hash function. used as checksum against transmission errors in opam.

**Comment by cfcs** [comment]
`and hope the md5 checksum is strong enough:` has been publicly known not to be strong enough for the better part of 10 years, see the [wikipedia article](https://en.wikipedia.org/wiki/MD5):
> A 2013 attack by Xie Tao, Fanbao Liu, and Dengguo Feng breaks MD5 collision resistance in 218 time. This attack runs in less than a second on a regular computer.

I believe the way to go about this would be implementing actual package signing so you wouldn't have to rely on neither `openssl` (notoriously bug-ridden) nor `md5` (most attackers can probably afford "less than a second" of CPU time on their 2013 "regular computer").

Fortunately it sounds like work is being done to support this, by implementing
a variant of [The Update Framework](http://theupdateframework.com/) for opam.
Perhaps @hannesm has more info on this process?
- [ocaml-x509 support for signing](https://github.com/mirleft/ocaml-x509/commit/380a645f7d08633f554dfcf1fb58a4380caa0391)

References to github issues related to this topic:
- https://github.com/ocaml/opam/issues/423
- https://github.com/ocaml/opam/issues/1370

**Comment by UnixJunkie** [comment]
opam packagers could sign their packages using GPG, as GNU asks you when you upload a tarball to them 

----- Mail original -----

**Comment by AltGr** [comment]
@UnixJunkie: yes, we know about this -- MD5 was used for checking the integrity of downloads and because it was readily available, not for matters of security. I and @hannesm have been working on a draft spec for a few weeks now and we'll open it for comments as soon as it feels consistent and well presented enough. It's indeed based on TUF and, @cfcs, your references at #423 have been very useful -- thanks.

We'll try to use stuff that's implemented in OCaml already (i.e. stuff in [nocrypto](http://opam.ocaml.org/packages/nocrypto/nocrypto.0.4.0/) and no GPG) -- who wants to rely on openssl anymore ? And end-to-end signing is only one part of the draft, see TUF for the attacks that it's not sufficient to prevent :)

**Comment by cfcs** [comment]
SO NICE to hear that this idea is gaining traction -- virtual pad on the shoulder to you two! :-)

GPG is indeed less than ideal in a lot of senses:
- history of vulnerabilities and overly complex code
- the OpenPGP format itself has a lot of problems
- not written in OCaml (and subject to entire classes memory corruption vulnerabilities that OCaml skips over): Hard to port to things like Mirage where verification of packages could come in handy
- the "web of trust" model is not well-suited for delegations of trust to specific domains; x509 allows much higher granularity in this respect. X509 has traditionally been avoided because of the complexity of implementation and because the "web of trust" model works relatively well for ensuring that your email goes to the right person (the original primary use case for GPG) - with `ocaml-x509` there is now such an implementation available in OCaml.

Really looking forward to seeing your work bear fruit!

**Comment by UnixJunkie** [comment]
@AltGr will the design use Merkle trees?

**Comment by AltGr** [comment]
No, but it's based in part on signing git commit-hashes, which results in something close (spoilers!) :)

**Comment by hannesm** [comment]
(see http://opam.ocaml.org/blog/Signing-the-opam-repository/ for a draft, feedback welcome)

**Comment by AltGr** [comment]
Closed by #2460
